abci/abci.go

```diff
@@ -90,126 +121,132 @@ func (h *ProposalHandler) ProcessProposalHandler() sdk.ProcessProposalHandler {
+		proposalInfo, partialProposals, err := h.ValidateBasic(req.Txs)
+		if err != nil {
+			h.logger.Error("failed to validate proposal", "err", err)
+			return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}, err
 		}
 
-		// Decode the transactions from the proposal.
-		decodedTxs, err := utils.GetDecodedTxs(h.txDecoder, txs)
+		// Build handler that will verify the partial proposals according to each lane's verification logic.
+		processLanesHandler := ChainProcessLanes(partialProposals, h.mempool.Registry())
+
+		// Build an empty placeholder proposal.
+		maxBlockSize, maxGasLimit := proposals.GetBlockLimits(ctx)
+		emptyProposal := proposals.NewProposal(h.txEncoder, maxBlockSize, maxGasLimit)
+
+		// Verify the proposal according to the verification logic from each lane.
+		finalProposal, err := processLanesHandler(ctx, emptyProposal)
 		if err != nil {
-			h.logger.Error("failed to decode transactions", "err", err)
+			h.logger.Error("failed to validate the proposal", "err", err)
 			return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}, err
 		}
 
-		// Verify the proposal using the verification logic from each lane.
-		if _, err := h.processLanesHandler(ctx, decodedTxs); err != nil {
+		// Ensure block size and gas limit are correct.
+		if err := h.ValidateBlockLimits(finalProposal, proposalInfo); err != nil {
 			h.logger.Error("failed to validate the proposal", "err", err)
 			return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}, err
 		}
 
-		h.logger.Info("validated proposal", "num_txs", len(txs))
+		h.logger.Info(
+			"processed proposal",
+			"num_txs", len(req.Txs),
+			"total_tx_bytes", finalProposal.Info.BlockSize,
+			"max_tx_bytes", maxBlockSize,
+			"total_gas_limit", finalProposal.Info.GasLimit,
+			"max_gas_limit", maxGasLimit,
+			"height", req.Height,
+		)
 
 		return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_ACCEPT}, nil
 	}
 }
 
-// ChainPrepareLanes chains together the proposal preparation logic from each lane
-// into a single function. The first lane in the chain is the first lane to be prepared and
-// the last lane in the chain is the last lane to be prepared.
-//
-// In the case where any of the lanes fail to prepare the partial proposal, the lane that failed
-// will be skipped and the next lane in the chain will be called to prepare the proposal.
-func ChainPrepareLanes(chain ...block.Lane) block.PrepareLanesHandler {
-	if len(chain) == 0 {
-		return nil
+// ValidateBasic validates the proposal against the basic invariants that are required
+// for the proposal to be valid. This includes:
+//  1. The proposal must contain the proposal information and must be valid.
+//  2. The proposal must contain the correct number of transactions for each lane.
+func (h *ProposalHandler) ValidateBasic(proposal [][]byte) (types.ProposalInfo, [][][]byte, error) {
```

Comment:
nit: The name of this fn is misleading, AFAICT `ValidateBasic` j returns in error (as in the cosmos-sdk), perhaps extract LaneProposals or smth